'''
This pipeline compares the absolute and relative expression of the genes in a
tissue to the rate at which target sites for a miRNA expressed in that tissue
undergo substitutions. This pipeline relies on 3 main datasets (and a few
subsidiary files):

1. A gene expression atlas conducted in 84 tissues (Su et al. 2004)
2. A miRNA expression atlas conducted in 172 tissues (Landgraf et al. 2007)
3. My observed and expected target site turnover results (run with the 'all_
together.py' wrapper script)
'''

import subprocess
data_folder='/home/alfred/Dropbox/Simkin_Grimson_final_ms_folder/figures_and_data/mammal_data/deeply_conserved_mirs(human_mouse_fish)/coexpression_data/'
transcript_exp_folder='human_gene_atlas/'
miRNA_exp_folder='miRNA_expression/'
simulation_folder='/mnt/md0/alfred/turnover_rates/4-28-15_deeper_phylogeny/simulations/'

miRNA_file=data_folder+miRNA_exp_folder+'human_w_random_precursor.tsv' #downloaded from Landgraf et al. 2007
conserved_miRNAs='/home/alfred/Dropbox/Simkin_Grimson_final_ms_folder/figures_and_data/mammal_data/deeply_conserved_mirs(human_mouse_fish)/main_results(vanilla)/mir_ranks_loss' #from all_together.py
mirbase_table='example_data_folder/mirbase_21_table_fixed.tsv' #from mirbase, manually curated a bit
affymetrix_file=data_folder+transcript_exp_folder+'affymetrix_annotation_10_7_14' #from biogps and GEO (repositories of Su et al. 2004 datasets)
gnf1h_file=data_folder+transcript_exp_folder+'gnf1h.annot2007.tsv' #from biogps and GEO (repositories of Su et al. 2004 datasets)
probe_file=data_folder+transcript_exp_folder+'U133AGNF1B.gcrma.avg.csv' #from biogps and GEO (repositories of Su et al. 2004 datasets)
alignment_file='/mnt/md0/alfred/turnover_rates/4-28-15_deeper_phylogeny/reformatted_anc_dictU' #generated by all_together.py
tissue_key=data_folder+'tissue_key' #created by user by comparing gene expression tissues with miRNA expression tissues
desired_smoothing='0.02' #paper used 0.02. Values near 1 create a lot of smoothing, values near 0 create very little

#scripts for calculating transcript expression:
#print 'using an annotation file to clean up and map probes to refseq genes from'
#print 'my alignments\n\n'
#subprocess.call(['python', 'map_probes_to_genes2.py', data_folder+transcript_exp_folder, affymetrix_file, gnf1h_file, probe_file, alignment_file])

#print '\n\nfiltering out multiply mapping probes, calculating geometric means per'
#print 'transcript, ranking relative and absolute expression of genes in tissue'
#subprocess.call(['python', 'relative_and_abs_exp.py', data_folder+transcript_exp_folder])

#print "\n\ncreating x and y coordinates for each transcript, where x is 'absolute'"
#print "expression and y is 'relative' expression"
#subprocess.call(['python', 'make_x_y.py', data_folder+transcript_exp_folder])


#scripts for calculating miRNA expression
#print '\n\ncalculating human miRNA expression levels using miRNA data from Landgraf'
#print 'et al. 2007'
#subprocess.call(['python', 'get_mir_exp3.py', data_folder+miRNA_exp_folder, conserved_miRNAs, mirbase_table])

#scripts for calculating coexpression
#print '\n\ngetting the tissue of highest expression for each miRNA, and x and y'
#print 'coordinates for each gene in that tissue (from make_x_y.py)'
#subprocess.call(['python', 'make_empty_mir_figure.py', data_folder])

#print "\n\ncounting turnover events for all miRNAs, getting 'expected' and"
#print "'observed' counts for each gene, and subtracting expected from observed"
#print 'to populate the coordinates from make_empty_mir_figure.py'
#subprocess.call(['python', 'populate_mir_figures.py', data_folder, simulation_folder, alignment_file])

print '\n\ngraphing the populated mir figures for the tissue of highest expression'
print 'of each miRNA'
subprocess.call(['python', 'graph_mir_figures.py', data_folder+'figures/', desired_smoothing])
#can also use graph_arbitrary_figure.py to graph any target site in any tissue
#of interest. If using this method, edit the 'miRNA' and 'shared_tissue_list'
#variables to hold the desired target site and tissues to examine
